File: src/server/lib/config-parser.test.ts
import { mkdirSync, mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { getConfigLayer, mergeConfigLayers } from "./config-parser.js";

describe("getConfigLayer", () => {
  let dir: string;

  beforeEach(() => {
    dir = mkdtempSync(join(tmpdir(), "config-parser-test-"));
  });

  afterEach(() => {
    rmSync(dir, { recursive: true, force: true });
  });

  it("reads a valid JSON file", () => {
    const filePath = join(dir, "settings.json");
    writeFileSync(filePath, JSON.stringify({ key: "value" }));

    const layer = getConfigLayer(filePath, "global");
    expect(layer.exists).toBe(true);
    expect(layer.content).toEqual({ key: "value" });
    expect(layer.source).toBe("global");
    expect(layer.filePath).toBe(filePath);
  });

  it("returns exists false for missing file", () => {
    const layer = getConfigLayer(join(dir, "nope.json"), "global");
    expect(layer.exists).toBe(false);
    expect(layer.content).toBeNull();
  });

  it("returns null content for invalid JSON", () => {
    const filePath = join(dir, "bad.json");
    writeFileSync(filePath, "not json{{{");

    const layer = getConfigLayer(filePath, "project");
    expect(layer.exists).toBe(true);
    expect(layer.content).toBeNull();
  });
});

describe("mergeConfigLayers", () => {
  let claudeHome: string;
  const originalEnv = process.env.CLAUDE_HOME;

  beforeEach(() => {
    claudeHome = mkdtempSync(join(tmpdir(), "claude-home-test-"));
    process.env.CLAUDE_HOME = claudeHome;
  });

  afterEach(() => {
    rmSync(claudeHome, { recursive: true, force: true });
    if (originalEnv === undefined) {
      delete process.env.CLAUDE_HOME;
    } else {
      process.env.CLAUDE_HOME = originalEnv;
    }
  });

  it("returns empty merged config when no files exist", () => {
    const result = mergeConfigLayers();
    expect(result.merged).toEqual({});
    expect(result.layers).toHaveLength(2);
  });

  it("reads global settings", () => {
    writeFileSync(join(claudeHome, "settings.json"), JSON.stringify({ theme: "dark" }));

    const result = mergeConfigLayers();
    expect(result.merged).toEqual({ theme: "dark" });
  });

  it("deep merges global and global-local", () => {
    writeFileSync(
      join(claudeHome, "settings.json"),
      JSON.stringify({ a: { x: 1, y: 2 }, b: "hello" }),
    );
    writeFileSync(
      join(claudeHome, "settings.local.json"),
      JSON.stringify({ a: { y: 99, z: 3 }, c: true }),
    );

    const result = mergeConfigLayers();
    expect(result.merged).toEqual({
      a: { x: 1, y: 99, z: 3 },
      b: "hello",
      c: true,
    });
  });

  it("includes project layers when projectPath is provided", () => {
    const projectDir = mkdtempSync(join(tmpdir(), "project-test-"));
    const claudeDir = join(projectDir, ".claude");
    mkdirSync(claudeDir);

    writeFileSync(join(claudeHome, "settings.json"), JSON.stringify({ global: true }));
    writeFileSync(join(claudeDir, "settings.json"), JSON.stringify({ project: true }));

    const result = mergeConfigLayers(projectDir);
    expect(result.layers).toHaveLength(4);
    expect(result.merged).toEqual({ global: true, project: true });

    rmSync(projectDir, { recursive: true, force: true });
  });

  it("later layers override earlier ones", () => {
    const projectDir = mkdtempSync(join(tmpdir(), "project-test-"));
    const claudeDir = join(projectDir, ".claude");
    mkdirSync(claudeDir);

    writeFileSync(join(claudeHome, "settings.json"), JSON.stringify({ mode: "global" }));
    writeFileSync(join(claudeDir, "settings.json"), JSON.stringify({ mode: "project" }));

    const result = mergeConfigLayers(projectDir);
    expect(result.merged.mode).toBe("project");

    rmSync(projectDir, { recursive: true, force: true });
  });

  it("arrays are replaced, not merged", () => {
    writeFileSync(join(claudeHome, "settings.json"), JSON.stringify({ items: [1, 2] }));
    writeFileSync(join(claudeHome, "settings.local.json"), JSON.stringify({ items: [3] }));

    const result = mergeConfigLayers();
    expect(result.merged.items).toEqual([3]);
  });
});

File: server/lib/configparser_test.go
package lib_test

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"fieldstation/lib"
)

func TestGetConfigLayer_ExistingFile(t *testing.T) {
	dir := t.TempDir()
	filePath := filepath.Join(dir, "settings.json")
	require.NoError(t, os.WriteFile(filePath, []byte(`{"key":"value"}`), 0o600))

	layer := lib.GetConfigLayer(filePath, lib.ConfigLayerGlobal)
	assert.True(t, layer.Exists)
	assert.Equal(t, lib.ConfigLayerGlobal, layer.Source)
	assert.Equal(t, filePath, layer.FilePath)
	assert.Equal(t, lib.JsonObject{"key": "value"}, layer.Content)
}

func TestGetConfigLayer_MissingFile(t *testing.T) {
	dir := t.TempDir()
	filePath := filepath.Join(dir, "nope.json")

	layer := lib.GetConfigLayer(filePath, lib.ConfigLayerGlobal)
	assert.False(t, layer.Exists)
	assert.Nil(t, layer.Content)
}

func TestGetConfigLayer_InvalidJSON(t *testing.T) {
	dir := t.TempDir()
	filePath := filepath.Join(dir, "bad.json")
	require.NoError(t, os.WriteFile(filePath, []byte("not json{{{"), 0o600))

	layer := lib.GetConfigLayer(filePath, lib.ConfigLayerProject)
	assert.True(t, layer.Exists)
	assert.Nil(t, layer.Content)
}

func TestMergeConfigLayers_EmptyNoFiles(t *testing.T) {
	claudeHome := t.TempDir()
	t.Setenv("CLAUDE_HOME", claudeHome)

	result := lib.MergeConfigLayers("")
	assert.Equal(t, lib.JsonObject{}, result.Merged)
	assert.Len(t, result.Layers, 2)
}

func TestMergeConfigLayers_GlobalSettings(t *testing.T) {
	claudeHome := t.TempDir()
	t.Setenv("CLAUDE_HOME", claudeHome)

	require.NoError(t, os.WriteFile(
		filepath.Join(claudeHome, "settings.json"),
		[]byte(`{"theme":"dark"}`),
		0o600,
	))

	result := lib.MergeConfigLayers("")
	assert.Equal(t, "dark", result.Merged["theme"])
}

func TestMergeConfigLayers_DeepMerge(t *testing.T) {
	claudeHome := t.TempDir()
	t.Setenv("CLAUDE_HOME", claudeHome)

	require.NoError(t, os.WriteFile(
		filepath.Join(claudeHome, "settings.json"),
		[]byte(`{"a":{"x":1,"y":2},"b":"hello"}`),
		0o600,
	))
	require.NoError(t, os.WriteFile(
		filepath.Join(claudeHome, "settings.local.json"),
		[]byte(`{"a":{"y":99,"z":3},"c":true}`),
		0o600,
	))

	result := lib.MergeConfigLayers("")
	a, ok := result.Merged["a"].(lib.JsonObject)
	require.True(t, ok)
	// x preserved, y overridden by local, z added from local
	assert.EqualValues(t, 1, a["x"])
	assert.EqualValues(t, 99, a["y"])
	assert.EqualValues(t, 3, a["z"])
	assert.Equal(t, "hello", result.Merged["b"])
	assert.Equal(t, true, result.Merged["c"])
}

func TestMergeConfigLayers_ProjectLayers(t *testing.T) {
	claudeHome := t.TempDir()
	t.Setenv("CLAUDE_HOME", claudeHome)

	projectDir := t.TempDir()
	claudeDir := filepath.Join(projectDir, ".claude")
	require.NoError(t, os.MkdirAll(claudeDir, 0o750))

	require.NoError(t, os.WriteFile(
		filepath.Join(claudeHome, "settings.json"),
		[]byte(`{"global":true}`),
		0o600,
	))
	require.NoError(t, os.WriteFile(
		filepath.Join(claudeDir, "settings.json"),
		[]byte(`{"project":true}`),
		0o600,
	))

	result := lib.MergeConfigLayers(projectDir)
	assert.Len(t, result.Layers, 4)
	assert.Equal(t, true, result.Merged["global"])
	assert.Equal(t, true, result.Merged["project"])
}

func TestMergeConfigLayers_LaterLayersOverride(t *testing.T) {
	claudeHome := t.TempDir()
	t.Setenv("CLAUDE_HOME", claudeHome)

	projectDir := t.TempDir()
	claudeDir := filepath.Join(projectDir, ".claude")
	require.NoError(t, os.MkdirAll(claudeDir, 0o750))

	require.NoError(t, os.WriteFile(
		filepath.Join(claudeHome, "settings.json"),
		[]byte(`{"mode":"global"}`),
		0o600,
	))
	require.NoError(t, os.WriteFile(
		filepath.Join(claudeDir, "settings.json"),
		[]byte(`{"mode":"project"}`),
		0o600,
	))

	result := lib.MergeConfigLayers(projectDir)
	assert.Equal(t, "project", result.Merged["mode"])
}

func TestMergeConfigLayers_ArraysReplaced(t *testing.T) {
	claudeHome := t.TempDir()
	t.Setenv("CLAUDE_HOME", claudeHome)

	require.NoError(t, os.WriteFile(
		filepath.Join(claudeHome, "settings.json"),
		[]byte(`{"items":[1,2]}`),
		0o600,
	))
	require.NoError(t, os.WriteFile(
		filepath.Join(claudeHome, "settings.local.json"),
		[]byte(`{"items":[3]}`),
		0o600,
	))

	result := lib.MergeConfigLayers("")
	assert.Equal(t, []any{float64(3)}, result.Merged["items"])
}
